File: app-runtime-management/agents/macos/Tests/MacOSAppRuntimeCoreTests/ApplicationClassificationTests.swift
import Foundation
import XCTest
@testable import MacOSAppRuntimeCore

final class ApplicationClassificationTests: XCTestCase {
    private struct Document: Decodable { let schemaVersion: Int; let cases: [Vector] }
    private struct Vector: Decodable {
        let name: String; let knowledge: ApplicationKnowledge; let childId: String
        let evidence: AppEvidence; let expected: AppClassificationResolution; let previous: String?
    }
    func testSharedVectors() throws {
        let package = URL(fileURLWithPath: #filePath).deletingLastPathComponent().deletingLastPathComponent().deletingLastPathComponent()
        let root = package.deletingLastPathComponent().deletingLastPathComponent()
        let url = root.appendingPathComponent("contracts/application-classification.vectors.json")
        let document = try JSONDecoder().decode(Document.self, from: Data(contentsOf: url))
        XCTAssertEqual(document.schemaVersion, 1)
        for vector in document.cases {
            XCTAssertEqual(ApplicationClassifier.resolve(vector.knowledge, childId: vector.childId,
                evidence: vector.evidence, previous: vector.previous ?? "unclassified"), vector.expected, vector.name)
        }
    }
}

File: app-runtime-management/agents/windows/tests/TimeOnChrome.AppRuntime.Core.Tests/ApplicationClassificationTests.cs
using System.Text.Json;
using TimeOnChrome.AppRuntime.Core;
using Xunit;

namespace TimeOnChrome.AppRuntime.Core.Tests;
public sealed class ApplicationClassificationTests
{
    [Fact]
    public void ConfirmedDistributionTypeRuleYieldsToExplicitClassification()
    {
        var product = new AppProduct("aimlabs","Aimlabs","game",[new("windows",new("all",[new("distributionKey","steam:714010")]))]);
        var rule = new AppClassificationRule("games","Games","type",new("all",[]),[],"automatic","restrictedEntertainment","game",true,"fixture","confirmed type");
        var evidence = new AppEvidence("windows","opaque","Aimlabs",new Dictionary<string,string>{{"distributionKey","steam:714010"}},["distributionKey"]);
        var automatic = ApplicationClassifier.Resolve(new(2,1,[product],[rule],[new("child",[],["games"])]),"child",evidence);
        Assert.Equal("game",automatic.AppType); Assert.Equal("confirmed",automatic.TypeStatus);
        Assert.Equal("restrictedEntertainment",automatic.Classification);
        var explicitResult = ApplicationClassifier.Resolve(new(2,1,[product],[rule],[new("child",[new("aimlabs","composite")],["games"])]),"child",evidence);
        Assert.Equal("game",explicitResult.AppType); Assert.Equal("composite",explicitResult.Classification);
    }
    [Fact]
    public void ReplaysSharedClassificationVectors()
    {
        var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
        var vectors = JsonSerializer.Deserialize<Vectors>(File.ReadAllText(
            Path.Combine(AppContext.BaseDirectory, "application-classification.vectors.json")), options)!;
        foreach (var vector in vectors.Cases)
        {
            var result = ApplicationClassifier.Resolve(vector.Knowledge, vector.ChildId,
                vector.Evidence, vector.Previous ?? "unclassified");
            Assert.Equal(vector.Expected.ProductId, result.ProductId);
            Assert.Equal(vector.Expected.Classification, result.Classification);
            Assert.Equal(vector.Expected.Status, result.Status);
            Assert.Equal(vector.Expected.RuleIds, result.RuleIds);
            Assert.Equal(vector.Expected.Suggestions, result.Suggestions);
        }
    }
    private sealed record Vectors(int SchemaVersion, IReadOnlyList<Vector> Cases);
    private sealed record Vector(string Name, ApplicationKnowledge Knowledge, string ChildId,
        AppEvidence Evidence, AppClassificationResolution Expected, string? Previous = null);
}
